pkg/services/store/object/object.proto

```diff
@@ -0,0 +1,242 @@
+package object;
+
+option go_package = "./;object";
+
+// Will be replaced with something from the SDK
+message UserInfo {
+  int64 id = 1; // internal grafana ID
+  string login = 2; // string ID? 
+}
+
+// Raw saved object metadata
+message RawObject {
+  // Unique ID
+  string UID = 1;
+
+  // Identify the object kind.  This kind will be used to apply a schema to the body and
+  // will trigger additional indexing behavior.
+  string kind = 2;
+
+  // Time in epoch milliseconds that the object was modified
+  int64 modified = 3;
+
+  // Who modified the object
+  UserInfo modified_by = 4;
+
+  // Content Length
+  int64 size = 5;
+
+  // MD5 digest of the Body
+  string ETag = 6;
+
+  // Raw bytes of the storage object.  The kind+schema will ensure
+  bytes body = 7;
+
+  // A small set of properties managed by raw storage layer.  The kind+schema will define val
+  map<string,string> properties = 8;
+
+  // object version (might be commit hash, or number)
+  string version = 9;
+
+  // optional "save" or "commit" message
+  string comment = 10;
+
+  // Time in epoch milliseconds that the object was last synced with an external system (provisioning/git)
+  int64 sync_time = 11;
+}
+
+// Report error while working with objects
+message ObjectErrorInfo {
+  // Match an error code registry?
+  int64 code = 1;
+
+  // Simple error display
+  string message = 2;
+
+  // Details encoded in JSON
+  bytes details_json = 3;
+}
+
+// Reference to 
+message ExternalReference {
+  // datasource, panel
+  string kind = 1;
+
+  // prometheus / heatmap
+  string type = 2;
+
+  // Unique ID for this object
+  string UID = 3;
+}
+
+
+//-----------------------------------------------
+// Get request/response
+//-----------------------------------------------
+
+message ReadObjectRequest {
+  string UID = 1;
+  string kind = 2;
+
+  // Fetch an explicit version
+  string version = 3;
+
+  // Include the full body bytes
+  bool with_body = 4;
+
+  // Include derived summary metadata
+  bool with_summary = 5;
+}
+
+message ReadObjectResponse {
+  // Object details with the body removed
+  RawObject object = 1; 
+
+  // Object summary as JSON
+  bytes summary_json = 2;
+}
+
+//-----------------------------------------------
+// Write request/response
+//-----------------------------------------------
+
+message WriteObjectRequest {
+  string UID = 1;
+  string kind = 2;
+
+  // The raw object body
+  bytes body = 3;
+
+  // Message that can be seen when exploring object history
+  string comment = 4;
+
+  // Raw properties from storage
+  map<string,string> properties = 5;
+
+  // Used for optimistic locking.  If missing, the previous version will be replaced regardless
+  string previous_version = 6;
+}
+
+message WriteObjectResponse {
+  // Error info -- if exists, the save did not happen
+  ObjectErrorInfo error = 1;
+
+  // Object details with the body removed
+  RawObject object = 2; 
+
+  // Object summary as JSON
+  bytes summary_json = 3;
```

Comment:
@ryantxu is the plan to add a field for a plugin to send extracted summary data with`WriteObjectRequest` for plugin defined kinds/types, or will grafana always be responsible for extracting that data?

Reply:
The summary should always be derived from the body.  It will not be sent in the `Write*` command.